### ISSUE TITLE
'Connection reset by peer - SSL_connect'

### ISSUE THREAD

**Issue by ctaymor** [cloudfoundry-incubator/bosh-azure-cpi-release#198]
When running bosh errands, we ran into an unknown CPI error. Unfortunately, we don't have the debug output because our CI rolled the bosh director before we could grab it. However, do you have any insights into what is going on with this SSL_connect failure?

It seems to be a flake, the next build ran successfully.

We're using version 14 of the CPI.

```
===== 2016-08-25 18:53:06 UTC Running "bundle exec bosh -n run errand example-errand"
Acting as client 'ops_manager' on deployment 'example-product-5a0a886abadb086c20f2' on 'p-bosh-aee2241a60542cf78619'
RSA 1024 bit CA certificates are loaded due to old openssl compatibility

Director task 5
  Started preparing deployment > Preparing deployment. Done (00:00:01)

  Started preparing package compilation > Finding packages to compile. Done (00:00:00)

  Started creating missing vms > example-errand/0 (cb99dab4-8fdb-4d3b-a4f4-55e9b96012b7). Done (00:01:48)

  Started updating job example-errand > example-errand/0 (cb99dab4-8fdb-4d3b-a4f4-55e9b96012b7) (canary). Done (00:00:42)

  Started running errand > example-errand/0. Done (00:00:01)

  Started fetching logs for example-errand/cb99dab4-8fdb-4d3b-a4f4-55e9b96012b7 (0) > Finding and packing log files. Done (00:00:01)

  Started deleting errand instances example-errand > example-errand/0 (cb99dab4-8fdb-4d3b-a4f4-55e9b96012b7). Failed: Unknown CPI error 'Unknown' with message 'Connection reset by peer - SSL_connect' (00:03:12)

Error 100: Unknown CPI error 'Unknown' with message 'Connection reset by peer - SSL_connect'

Task 5 error

Errand 'example-errand' did not complete

For a more detailed error report, run: bosh task 5 --debug
===== 2016-08-25 18:58:52 UTC Finished "bundle exec bosh -n run errand example-errand"; Duration: 345s; 
```

**Comment by AbelHu** [comment]
@ctaymor It seems like that the task failed in deleting a disk. Azure storage closed the https connection when CPI was calling delete_blob of azure-storage-ruby.
We cannot investigate it since you cannot provide debug logs. But we will consider to log more details when an exception occurs in azure-storage-ruby.

**Comment by ctaymor** [comment]
Thanks @AbelHu. If we see it again, we'll grab the debug logs and share them.

**Comment by cppforlife** [comment]
@abelhu i think we shpuld retry all http requests done by cpi *specifically on ssl_connect errors*. we do this in bosh cli as well since connect errors are ok retry blindly.

**Comment by AbelHu** [comment]
@cppforlife Thanks for the suggestion. We are asking SDK team to add retry logic to handle this.

**Comment by AbelHu** [comment]
Close it since azure-storage-ruby has fixed this in v0.11.1 and Azure CPI has upgraded the SDK to v0.11.1 in this [commit](https://github.com/cloudfoundry-incubator/bosh-azure-cpi-release/commit/90be7c11f2ede00136e177873a651e96d018c981).
